src/mscorlib/shared/System/Memory.cs

```diff
@@ -0,0 +1,287 @@
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <param name="start">The index at which to begin the memory.</param>
+        /// <param name="length">The number of items in the memory.</param>
+        /// <exception cref="System.ArgumentNullException">Thrown when <paramref name="array"/> is a null
+        /// reference (Nothing in Visual Basic).</exception>
+        /// <exception cref="System.ArrayTypeMismatchException">Thrown when <paramref name="array"/> is covariant and array's type is not exactly T[].</exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> or end index is not in the range (&lt;0 or &gt;=Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Memory(T[] array, int start, int length)
+        {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            if (default(T) == null && array.GetType() != typeof(T[]))
+                ThrowHelper.ThrowArrayTypeMismatchException();
+            if ((uint)start > (uint)array.Length || (uint)length > (uint)(array.Length - start))
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            _arrayOrOwnedMemory = array;
+            _index = start;
+            _length = length;
+        }
+        
+        // Constructor for internal use only.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Memory(OwnedMemory<T> owner, int index, int length)
+        {
+            if (owner == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.ownedMemory);
+            if (index < 0 || length < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            _arrayOrOwnedMemory = owner;
+            _index = index | (1 << 31); // Before using _index, check if _index < 0, then 'and' it with RemoveOwnedFlagBitMask
+            _length = length;
+        }
+
+        /// <summary>
+        /// Defines an implicit conversion of an array to a <see cref="Memory{T}"/>
+        /// </summary>
+        public static implicit operator Memory<T>(T[] array) => new Memory<T>(array);
+        
+        /// <summary>
+        /// Defines an implicit conversion of a <see cref="ArraySegment{T}"/> to a <see cref="Memory{T}"/>
+        /// </summary>
+        public static implicit operator Memory<T>(ArraySegment<T> arraySegment) => new Memory<T>(arraySegment.Array, arraySegment.Offset, arraySegment.Count);
+
+        /// <summary>
+        /// Defines an implicit conversion of a <see cref="Memory{T}"/> to a <see cref="ReadOnlyMemory{T}"/>
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static implicit operator ReadOnlyMemory<T>(Memory<T> memory)
+        {
+            if (memory._index < 0)
```

Comment:
Did you experiment with doing 

```csharp
var arrayOrOwnedMemory = _arrayOrOwnedMemory;
if (arrayOrOwnedMemory.GetType() != typeof(T[])) 
    return new ReadOnlyMemory<T>(Unsafe.As<OwnedMemory<T>>(arrayOrOwnedMemory), memory._index & RemoveOwnedFlagBitMask, memory._length); 
return new ReadOnlyMemory<T>(Unsafe.As<T[]>(arrayOrOwnedMemory), memory._index, memory._length); 
```
rather than the sign of index trick?  This would give type safety.

Reply:
This does not work when the array type is not exact. E.g.:

```
enum X {
}

    X[] a = new X[100];
    uint[] b = (uint[])(object)a;
    Console.WriteLine(b.GetType());
```

Reply:
I thought you were ruling out Covariant typing on Memory and Span? Or do I not understand your example.

Reply:
Sorry I hadn't spotted the 
```
default(T) == null 
```
in the constructor.  I guess this means you could create a Span of type "uint" where the underlying array is of type "X".  Why do you need/want this condition?

Reply:
> Covariant typing

For references, but not for primitive valuetypes - in the current implementation at least.

I am not sure whether covariance is the right name for this behavior of primitive valuetypes. Covariance is asymmetric, but the relationship between `uint[]` and `X[]` in the above example is symmetric.

Reply:
> Why do you need/want this condition?

This condition is there to eliminate the type check for valuetypes. If we remove it, creating Span from `byte[]` will be slower. We do not want creating Span from `byte[]` to be slower.

Reply:
That applies to `Span`, but it doesn't apply to `Memory`, e.g. the `Span` property has a cast inside:
```csharp
return new Span<T>((T[])_arrayOrOwnedMemory, _index, _length); 
```
However, I can see you want them to be consistent.

Are there cases where `byte[]` can be viewed at a different type? If not, the JIT could optimize GetType on anything that is forced to be a single type.

Reply:
> Are there cases where byte[] can be viewed at a different type

Yes.

```
enum X : byte {
}


    X[] a = new X[100];
    byte[] b = (byte[])(object)a;
    Console.WriteLine(b.GetType());
```